src/lib/SaferNames/Syntax.hs

```diff
@@ -154,135 +157,62 @@ type Con n = PrimCon (Atom n)
 pattern Pure <- ((\(EffectRow effs t) -> (S.null effs, t)) -> (True, Nothing))
  where  Pure = EffectRow mempty Nothing
 
--- === Recursive bindings ===
-
-data BinderInfo n =
-   LamBound (ArrowP ())
- -- TODO: make the expression optional, for when it's effectful?
- -- (or we could put the effect tag on the let annotation)
- | PatBound
- | LetBound LetAnn (Expr n)
- | PiBound
- | UnknownBinder
- | DataDefName (DataDef n)
- | ClassName -- TODO
- | MethodName  (NamedDataDef n) Int  -- TODO: just have a class name and a label
- | DataConName (NamedDataDef n) Int
- | TypeConName (NamedDataDef n)
- -- TODO: avoid this case by having the inference monad use an alternative
- -- version of BinderInfo
- | InferenceName
-
--- The `Nothing` case is for things without types, like `DataDefName`. Everything
--- that can have a type must have a type.
-data TypedBinderInfo n = TypedBinderInfo (Maybe (Type n)) (BinderInfo n)
-
-type Bindings     = RecEnv     TypedBinderInfo
-type BindingsFrag = RecEnvFrag TypedBinderInfo
-
--- === traversals with atom substitutions ===
-
-data SubstVal e n = SubstVal (e n) | Rename (Name n)
-
--- TODO: make this just `Env (Atom o) i` if we remove type annotations from
--- variable occurrences
-type AtomSubstVal = SubstVal Atom
-type SubstEnv i o = Env i (AtomSubstVal o)
-
-type CoreTraversal = NameTraversal AtomSubstVal
-
-class HasNamesE e => SubstE e where
-  traverseCoreE
-    :: Monad m
-    => Scope o
-    -> CoreTraversal m i o
-    -> e i
-    -> m (e o)
-
-class HasNamesB b => SubstB (b::B) where
-  traverseCoreB
-    :: Monad m
-    => Scope o
-    -> CoreTraversal m i o
-    -> b i i'
-    -> m (FreshBinder b o (i:=>:i') )
-
-  asBindingsFrag :: b n l -> BindingsFrag n l
-
-coreTraversalFromRenameTraversal :: Monad m => RenameTraversal m i o -> CoreTraversal m i o
-coreTraversalFromRenameTraversal (NameTraversal f renamer) =
-  NameTraversal f' renamer
-  where f' name = Rename <$> f name
-
-traverseNamesFromSubstE :: (SubstE e, Monad m)
-                           => Scope o -> RenameTraversal m i o -> e i -> m (e o)
-traverseNamesFromSubstE scope t e =
-  traverseCoreE scope (coreTraversalFromRenameTraversal t) e
-
-traverseNamesFromSubstB
-  :: (SubstB b, Monad m)
-  => Scope o -> RenameTraversal m i o -> b i i' -> m (FreshBinder b o (i:=>:i'))
-traverseNamesFromSubstB s t b =
-  traverseCoreB s (coreTraversalFromRenameTraversal t) b
-
 -- === Synonyms ===
 
-binderType :: Binder n l -> Type n
-binderType (_:>ann) = ann
-
 infixr 1 -->
 infixr 1 --@
 infixr 2 ==>
 
-(-->) :: Type n -> Type n -> Type n
-a --> b = Pi (Abs (Ignore:>a) (WithArrow PureArrow b))
+nonDepPiType :: ScopeReader m => Arrow -> Type n -> EffectRow n -> Type n -> m n (Type n)
+nonDepPiType arr argTy eff resultTy =
+  toConstAbs (TypedBinderInfo argTy PiBound) (PairE eff resultTy) >>= \case
+    Abs b (PairE eff' resultTy') ->
+      return $ Pi $ PiType arr (b:>argTy) eff' resultTy'
+
+(?-->) :: ScopeReader m => Type n -> Type n -> m n (Type n)
```

Comment:
Why are those functions monadic now? It seems to me that if you have two types in the same scope and want to build a non-dependent lambda then you should just be able to do it. Is it because we no longer have the `Ignore` binders and we have to use the scope to come up with a fresh name?

Reply:
Yes, exactly. It's the price we pay for dropping `Ignore`. I wrote about it a bit in the commit message to af82c443b5709f3677ef05efbc14a64cdb48bbea. We can always make a "scopeless" variant that queries the names in the types to create a fresh one, but that has performance implications.